common/defs.go

```diff
@@ -197,58 +199,74 @@ func (p *Policy) Validate() error {
+				errMsg = append(errMsg, fmt.Sprintf("Rule #%d: The following ports are invalid: %s.", ruleNo, strings.Join(badPorts, ", ")))
 			}
 		}
-		if r.Protocol == "icmp" {
+		if r.Protocol != "icmp" {
 			if r.IcmpCode > 0 || r.IcmpType > 0 {
-				errMsg += "ICMP protocol is not specified but ICMP Code and/or ICMP Type are also specified. "
-			} else {
-				if len(r.Ports) > 0 || len(r.PortRanges) > 0 {
-					errMsg += "ICMP protocol is specified but ports are also specified. "
+				errMsg = append(errMsg, fmt.Sprintf("Rule #%d: ICMP protocol is not specified but ICMP Code and/or ICMP Type are also specified.", ruleNo))
+			}
+		} else {
+			if len(r.Ports) > 0 || len(r.PortRanges) > 0 {
+				errMsg = append(errMsg, fmt.Sprintf("Rule #%d: ICMP protocol is specified but ports are also specified.", ruleNo))
+			}
+			if r.IcmpType < 0 || r.IcmpType > MaxIcmpType {
+				errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid ICMP type: %d.", ruleNo, r.IcmpType))
+			}
+			switch r.IcmpType {
+			case 3: // Destination unreachable
+				if r.IcmpCode < 0 || r.IcmpCode > 15 {
+					errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid ICMP code for type %d: %d.", ruleNo, r.IcmpType, r.IcmpCode))
 				}
-				if r.IcmpType < 0 || r.IcmpType > MaxIcmpType {
-					errMsg += fmt.Sprintf("Invalid ICMP type: %d. ", r.IcmpType)
+			case 4: // Source quench
+				if r.IcmpCode != 0 {
+					errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid ICMP code for type %d: %d.", ruleNo, r.IcmpType, r.IcmpCode))
 				}
-				switch r.IcmpType {
-				case 3: // Destination unreachable
-					if r.IcmpCode < 0 || r.IcmpCode > 15 {
-						errMsg += fmt.Sprintf("Invalid ICMP code for type %d: %d", r.IcmpType, r.IcmpCode)
-					}
-				case 4: // Source quench
-					if r.IcmpCode != 0 {
-						errMsg += fmt.Sprintf("Invalid ICMP code for type %d: %d", r.IcmpType, r.IcmpCode)
-					}
-				case 5: // Redirect
-					if r.IcmpCode < 0 || r.IcmpCode > 3 {
-						errMsg += fmt.Sprintf("Invalid ICMP code for type %d: %d", r.IcmpType, r.IcmpCode)
-					}
-				case 11: // Time exceeded
-					if r.IcmpCode < 0 || r.IcmpCode > 1 {
-						errMsg += fmt.Sprintf("Invalid ICMP code for type %d: %d", r.IcmpType, r.IcmpCode)
-					}
-				default:
-					if r.IcmpCode != 0 {
-						errMsg += fmt.Sprintf("Invalid ICMP code for type %d: %d", r.IcmpType, r.IcmpCode)
-					}
-
+			case 5: // Redirect
+				if r.IcmpCode < 0 || r.IcmpCode > 3 {
+					errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid ICMP code for type %d: %d.", ruleNo, r.IcmpType, r.IcmpCode))
+				}
+			case 11: // Time exceeded
+				if r.IcmpCode < 0 || r.IcmpCode > 1 {
+					errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid ICMP code for type %d: %d.", ruleNo, r.IcmpType, r.IcmpCode))
 				}
+			default:
+				if r.IcmpCode != 0 {
+					errMsg = append(errMsg, fmt.Sprintf("Rule #%d: Invalid ICMP code for type %d: %d.", ruleNo, r.IcmpType, r.IcmpCode))
+				}
+
 			}
+
+		}
+	}
+	for i, endpoint := range p.AppliedTo {
```

Comment:
I think we could at least provide a docstring as heading for this new chapter here. We just finished the rules, now we move on to processing the 'applied-to' section. We should have some comments at the start of each of those major sections.